pthread/pthread.c

```diff
@@ -1171,6 +1204,72 @@ void _pthread_atfork_child(void)
 }
 
 
+void pthread_cleanup_push(void (*routine)(void *), void *arg)
+{
+	pthread_ctx *ctx = find_pthread(gettid());
+
+	if (ctx == NULL) {
+		return;
+	}
+
+	mutexLock(pthread_common.pthread_list_lock);
+
+	if (ctx->cleanup_list == NULL) {
+		ctx->cleanup_list = (pthread_cleanup_t *)malloc(sizeof(pthread_cleanup_t));
+		if (ctx->cleanup_list == NULL) {
+			_pthread_ctx_put(ctx);
+			return;
+		}
+		ctx->cleanup_list->next = NULL;
+	}
+	else {
+		pthread_cleanup_t *head = (pthread_cleanup_t *)malloc(sizeof(pthread_cleanup_t));
+		if (head == NULL) {
+			_pthread_ctx_put(ctx);
+			return;
+		}
+		head->next = ctx->cleanup_list;
+		ctx->cleanup_list = head;
+	}
+
+	ctx->cleanup_list->routine = routine;
+	ctx->cleanup_list->arg = arg;
+
+	_pthread_ctx_put(ctx);
+}
+
+
+void pthread_cleanup_pop(int execute)
+{
+	pthread_ctx *ctx = find_pthread(gettid());
+
+	if (ctx == NULL) {
+		return;
+	}
+
+	mutexLock(pthread_common.pthread_list_lock);
+
+	if (ctx->cleanup_list == NULL) {
+		_pthread_ctx_put(ctx);
+		return;
+	}
+
+	pthread_cleanup_t *head = ctx->cleanup_list;
+	ctx->cleanup_list = head->next;
+	_pthread_ctx_put(ctx);
+
+	if (execute != 0) {
+		void (*routine)(void *) = head->routine;
+		void *arg = head->arg;
+		free(head);
+		routine(arg);
+	}
+	else {
+		free(head);
+	}
```

Comment:
```suggestion
		routine(arg);
	}

	free(head);
```

Reply:
Or is it essential to `free(head)` before `routine()`? Can we e.g. die in the `routine()`?

Reply:
I'd leave it as it is, nothing in POSIX prevents us from calling e.g. `pthread_exit` from cleanup routine, so we'd end up with unfreed memory